src/stream/src/executor/source/source_executor.rs

```diff
@@ -610,6 +653,58 @@ impl<S: StateStore> Debug for SourceExecutor<S> {
     }
 }
 
+struct WaitEpochWoker<S: StateStore> {
+    wait_epoch_rx: UnboundedReceiver<(Epoch, HashMap<SplitId, String>)>,
+    state_table_handler: Arc<Mutex<SourceStateTableHandler<S>>>,
+}
+
+impl<S: StateStore> WaitEpochWoker<S> {
+    pub async fn run(mut self) {
+        tracing::debug!("wait epoch worker start success");
+        loop {
+            // poll the rx and wait for the epoch commit
+            match self.wait_epoch_rx.recv().await {
+                Some((epoch, updated_offsets)) => {
+                    let state_store_guard = self.state_table_handler.lock().await;
+                    tracing::debug!("start to wait epoch {}", epoch.0);
+                    let ret = state_store_guard.state_store.wait_epoch(epoch.0).await;
```

Comment:
The calling of `wait_epoch` resides in the critical section of the lock over `state_store_handler`. This looks unnecessary to me and could lead to deadlocks.